compiler/modules/bir/verify.bal

```diff
@@ -79,6 +80,54 @@ type Error err:Semantic|err:Internal;
 function verifyBasicBlock(VerifyContext vc, BasicBlock bb) returns Error? {
     foreach Insn insn in bb.insns {
         check verifyInsn(vc, insn);
+        check verifyRegistersKinds(vc, insn);
+    }
+}
+
+function verifyNonFinalRegisterKind(VerifyContext vc, Operand r) returns Error? {
+    if r is FinalRegister {
+            return vc.invalidErr("invalid register kind final for insn: " + r.kind, <Position>r.pos);
+    }
+}
+ 
+function verifyRegisterKind(VerifyContext vc, Operand r) returns Error? {
+    if r !is Register {
+        return;
+    }
+    if r is TmpRegister {
+        if vc.usedTmpRegisters.indexOf(r.number, 0) != () {
+            return vc.invalidErr("tmp register used in multiple places", <Position>r.pos);
+        }
+        else {
+            vc.usedTmpRegisters.push(r.number);
+        }
+    }
+}
+
+type MultipleOpeandInsn IntBinaryInsn|IntNoPanicArithmeticBinaryInsn|FloatArithmeticBinaryInsn
+    |DecimalArithmeticBinaryInsn|CompareInsn|ListConstructInsn|ListGetInsn|MappingConstructInsn
+    |MappingGetInsn|StringConcatInsn;
+
+
+function verifyRegistersKinds(VerifyContext vc, Insn insn) returns Error? {
+    if insn is BranchInsn|CatchInsn|EqualityInsn|AssignInsn|CondBranchInsn|CondNarrowInsn {
+        return;
+    }
+    else if insn is CallInsn {
+        foreach Operand arg in insn.args {
+            check verifyRegisterKind(vc, arg);
+        }
+    }
+    else if insn is ListSetInsn|MappingSetInsn {
+        check verifyNonFinalRegisterKind(vc, insn.operands[0]);
+    }
+    else if insn is MultipleOpeandInsn {
+        foreach var op in <Operand[]>insn.operands {
```

Comment:
@jclark without the type cast it throws the error 
`ERROR [verify.bal:(125:27,125:40)] incompatible types: '((wso2/nballerina.bir:0.1.0:IntOperand[2] & readonly)|(wso2/nballerina.bir:0.1.0:IntOperand[2] & readonly)|(wso2/nballerina.bir:0.1.0:IntOperand[2] & readonly)|(wso2/nballerina.bir:0.1.0:FloatOperand[2] & readonly)|(wso2/nballerina.bir:0.1.0:DecimalOperand[2] & readonly)|(wso2/nballerina.bir:0.1.0:Operand[2] & readonly)|(wso2/nballerina.bir:0.1.0:Operand[] & readonly)|([wso2/nballerina.bir:0.1.0:Register,wso2/nballerina.bir:0.1.0:IntOperand] & readonly)|(wso2/nballerina.bir:0.1.0:Operand[] & readonly)|([wso2/nballerina.bir:0.1.0:Register,wso2/nballerina.bir:0.1.0:StringOperand] & readonly)|(wso2/nballerina.bir:0.1.0:StringOperand[2] & readonly))' is not an iterable collection`

This seems independent of the insn type. Even if we use just 2 instructions with same `operands` it still fails.
For `type MultipleOpeandInsn ListConstructInsn|MappingConstructInsn;` 
`incompatible types: '((wso2/nballerina.bir:0.1.0:Operand[] & readonly)|(wso2/nballerina.bir:0.1.0:Operand[] & readonly))' is not an iterable collection`

Reply:
Added an issue for JBUG in https://github.com/ballerina-platform/ballerina-lang/issues/35557

Reply:
Can we work around by assigning to a variable before iterating?

Reply:
Yes, changed it to assign and iterate.